Add new masterclass content

mforner13 opened the pull request:
This PR adds the first part of the 'Create a masterclass' journey. It allows users to choose the content of their masterclass, i.e. a related DDaT job family, a name and a description. 

When a user clicks 'Start now' on the start page, a draft masterclass is created with just an id. This id is stored in the session and accessed in order to add either existing masterclass content to the draft masterclass, or new masterclass content. This approach was chosen over storing data from each section of the 'Create a masterclass' journey in the session and creating a new masterclass at the end of the journey, so as to avoid the user losing data they had already entered. 

The draft masterclass is created by clicking start now, rather than before or after this point, so as to both reduce the number of incomplete draft masterclasses in the databases and to avoid a user losing too much of their data by doing it later on in the process.

**TODO**
Another PR is to follow which updates the logged_in_user fixture to actually be logged in, as well as additional tests to check that a user cannot access certain routes unless they are logged in.

**Screenshots**
1. Start create masterclass journey
![Screenshot 2020-06-29 at 21 50 05](https://user-images.githubusercontent.com/22460823/86054683-80403e80-ba52-11ea-9159-9dceaef8dd3d.png)

2. Choose a DDaT job family
![Screenshot 2020-06-29 at 21 50 27](https://user-images.githubusercontent.com/22460823/86054715-8df5c400-ba52-11ea-8440-d7f7e536de2d.png)

3. Chose existing masterclass content or create new
![Screenshot 2020-06-29 at 21 50 49](https://user-images.githubusercontent.com/22460823/86054751-9b12b300-ba52-11ea-9833-71d925ea1ab5.png)

4. Create new masterclass content
<img width="737" alt="Screenshot 2020-07-01 at 22 07 40" src="https://user-images.githubusercontent.com/22460823/86291697-49e7f800-bbe7-11ea-8ee3-b95d1869b726.png">

jonodrew commented on app/templates/create-masterclass/content/choose-ddat-family.html:
```diff
@@ -0,0 +1,71 @@
+                  Which DDaT job family is your masterclass related to?
+                </h1>
+              </legend>
+              <span id="sign-in-hint" class="govuk-hint">
```
What's this for, and what's it doing here?

mforner13 replied:
The span adds the spacing below the heading. I must have just kept the id from copying and pasting from the design system :p Have corrected.

jonodrew commented on app/templates/create-masterclass/content/new-or-existing.html:
```diff
@@ -0,0 +1,52 @@
+                  Which masterclass do you want to run?
+                </h1>
+              </legend>
+              <span id="sign-in-hint" class="govuk-hint">
```
Here it is again - why is this a `span` with an id of `sign-in-hint` ?

jonodrew commented on app/templates/create-masterclass/content/new-or-existing.html:
```diff
@@ -0,0 +1,52 @@
+                {% for masterclass in existing_masterclasses %}
+                    <div class="govuk-radios__item">
+                        <input class="govuk-radios__input" id="existing-masterclass" name="which-masterclass" type="radio" value={{masterclass.name}}>
+                        {# How do I add index of masterclass in list of existing to id? {{exiting_masterclasses.index(masterclass)}} ? #}
```
In general we try not to leave open questions in our code :laughing:

jonodrew commented on tests/test_masterclasses.py:
```diff
@@ -4,39 +4,54 @@
-    assert rv.status_code == 302
+def test_login_required(logged_in_user, test_client):
+    rv = test_client.get('/')
+    assert rv.status_code == 200
```
What do these tests do?

mforner13 replied:
Agreed, I have removed them and left a more detailed commit message explaining why.

jonodrew commented on app/templates/index.html:
```diff
@@ -5,7 +5,7 @@
         <h1 class="govuk-heading-l">Upcoming masterclasses</h1>
 
-        {% for masterclass in masterclasses %}
+        {% for masterclass in masterclasses if not masterclass.draft %}
```
Why do you filter in the template and not on the server?

mforner13 replied:
So I couldn't find a way of doing it in SQLAlchemy so got all Masterclasses and then used a for loop. Have I missed something obvious?

```
def index():
    all_masterclasses = Masterclass.query.order_by(Masterclass.timestamp.asc()).all()
    masterclasses = []
    for m in all_masterclasses:
        if not m.draft:
            masterclasses.append(m)
    return render_template('index.html', title='Home', user=User, masterclasses=masterclasses)
```

jonodrew replied:
Could you use `filter_by(Masterclass.draft is None)` in your query?

jonodrew replied:
And, if not, there's a neat one-liner for your filter - you can compress four lines down to
`masterclasses = [masterclass for masterclass in all_masterclasses if masterclass.draft is not None]`

mforner13 replied:
The `filter_by(draft=None)` worked 👍

jonodrew commented on app/routes.py:
```diff
@@ -106,9 +106,14 @@ def choose_new_or_existing_content():
-            db.session.add(new_masterclass_content) 
+            new_content = MasterclassContent(name = request.form['masterclass-name'], description = request.form['masterclass-description'])
+            db.session.add(new_content) 
             db.session.commit()
```
Is there any way we can make this one db call? Do we need to commit the object to the db to get an `id`? (Genuinely don't know, interested in the answer!)

mforner13 replied:
Unfortunately an ID isn't created until the object is committed to the database, so in order to assign that ID to the draft masterclass, it needs to be committed.

jonodrew commented on tests/test_masterclasses.py:
```diff
@@ -109,3 +109,12 @@ def test_chosen_job_family_is_added_to_session(test_client, logged_in_user):
+    with test_client.session_transaction() as session:    
+        session['draft_masterclass_id'] = 1
+        with test_client.post('/choose-content/create-new', data = {'masterclass-name': 'Test name', 'masterclass-description': 'A description'}):
+            draft_masterclass = Masterclass.query.filter_by(id=1).first()
```
I think `.first()` is unnecessary here? Can you use `.get(id)` ?

jonodrew commented on tests/test_masterclasses.py:
```diff
@@ -109,3 +109,12 @@ def test_chosen_job_family_is_added_to_session(test_client, logged_in_user):
+            draft_masterclass = Masterclass.query.filter_by(id=1).first()
+            new_content = MasterclassContent.query.filter_by(name='Test name').first()
+            assert draft_masterclass.masterclass_content_id == new_content.id
+            # Need to add draft masterclass id to the session first!
```
Love the comments for me, but they shouldn't make it into your commit history

jonodrew commented on app/routes.py:
```diff
@@ -109,9 +113,8 @@ def create_new_content():
-            new_content_id = new_content.id
             draft_masterclass = Masterclass.query.filter_by(id=session['draft_masterclass_id']).first() 
-            draft_masterclass.masterclass_content_id = new_content_id
+            draft_masterclass.masterclass_content_id = new_content.id
```
Nice! :+1:

jonodrew commented on tests/test_masterclasses.py:
```diff
@@ -94,27 +117,36 @@ def test_user_is_displayed_the_correct_template_according_to_request(test_app, t
\ No newline at end of file
+    masterclass_content = MasterclassContent.query.filter_by(name='Introduction to R').first()
+    logged_in_user.post('/choose-content/new-or-existing', data = {'which-masterclass': masterclass_content.id})
+    draft_masterclass = Masterclass.query.filter_by(id=1).first()
```
Same as before - and in fact how you've done it in line 142

jonodrew reviewed (changes requested):
Can you review your commit messages as well as my feedback and give them a bit more meat. Otherwise very happy with this :+1:

mforner13 commented on app/templates/create-masterclass/content/create-new.html:
```diff
@@ -0,0 +1,80 @@
+                <textarea class="govuk-textarea" id="masterclass-description" name="masterclass-description" rows="5" aria-describedby="masterclass-description"></textarea>
+                <div class="govuk-radios">
+                    <label class="govuk-label" for="sort">
+                      Which DDaT 
```
Should be: Which DDaT job family is your masterclass related to?

mforner13 replied:
Could extract DDaT job family radio buttons into separate template to make more DRY.

mforner13 replied:
Made into a partial template.

mforner13 committed:
Remove login required tests

I am removing these tests as they do not currently properly test the behaviour of a not logged in user. My next PR will include more fleshed out versions of these tests and additional ones for other routes.

mforner13 committed:
Fix skewed indentation

mforner13 committed:
Add init file for test app

In order for the test app to run there needs to be an __init__.py file in the test folder. This can remain blank.

mforner13 committed:
Don't display draft masterclass on homepage

mforner13 committed:
Add start page and create draft masterclass

This commit also includes a test which checks that a draft masterclass is created upon posting to create-masterclass, and that its id is added to the session. As part of this I have separated the test_app and the test_client, created a logged_in_user fixture and created a blank session fixture which will allow for this and future tests to interact with the db session.

mforner13 committed:
Choose ddat job family and add choice to session

The chosen job family is added to the session so that when a user goes back to the next step in the form, in which they're displayed the content items related to that job family, we have access to what option they originally chose. The job family radio buttons are extracted into their own template as they will be reused when a user adds new masterclass content.

jonodrew commented:
I've seen you've been resolving things but I've not yet seen any new code - would you mind requesting a new review once you've pushed up the changes please?

mforner13 committed:
Display existing content and add to draft on post

Existing content for the job family in the session is displayed. If a user chooses an existing one, the id for that item is added to the draft masterclass, if not a user is directed to the 'create new' page. I have included a test which checks if the content id is successfully added to the draft masterclass.

mforner13 committed:
Add new masterclass content

Users can create new content for their masterclass, this is committed to the database and then its id is added to the draft masterclass.

mforner13 committed:
Add parametrised tests for content routes

I have added two tests, one that checks that the correct template is rendered, and one that checks the user is redirected correctly depending on whether they choose new or existing content.

mforner13 committed:
Install blinker

Blinker allows the use of signals in the captured templates fixture.

mforner13 commented:
@jonodrew Thanks for the comments! I have re-written the commit history to hopefully be much more clear and succinct, as well as made the suggested changes. Let me know what you think!

jonodrew commented on app/templates/create-masterclass/content/choose-ddat-family.html:
```diff
@@ -0,0 +1,34 @@
+                </h1>
+              </legend>
+              <span id="DDaT-job-family-link" class="govuk-hint">
+                <a href="https://www.gov.uk/government/collections/digital-data-and-technology-profession-capability-framework#data-job-family">Find out more about the job families</a>
```
Should this link be more general, or the text be more specific? 
```suggestion
                <a href="https://www.gov.uk/government/collections/digital-data-and-technology-profession-capability-framework">Find out more about the job families</a>
```

mforner13 replied:
I think text should be more specific, so have edited the text to be `Find out more about the DDaT job families`.

jonodrew commented on app/routes.py:
```diff
@@ -109,3 +109,18 @@ def choose_new_or_existing_content():
+            new_content = MasterclassContent(name = request.form['masterclass-name'], description = request.form['masterclass-description'])
+            db.session.add(new_content) 
+            db.session.commit()
+            draft_masterclass = Masterclass.query.filter_by(id=session['draft_masterclass_id']).first() 
```
Could this be another `.get`?
```suggestion
            draft_masterclass = Masterclass.query.get(session['draft_masterclass_id']) 
```

jonodrew reviewed (changes requested):
A couple of small things here to fixup and then it's ready to rock. Really awesome work Miriam!

mforner13 committed:
Update DDaT job family link text

mforner13 committed:
Simplify query to get masterclass in sesison

jonodrew reviewed (approved):
Really fantastic work Miriam, looks great!
